packages/nimble-components/src/rich-text-mention/base/models/mention-validator.ts

```diff
@@ -0,0 +1,147 @@
+        this.validateUniqueMentionHref(mentionHrefs);
+        this.validateMissingPattern(pattern);
+        this.validatePattern(pattern);
+        this.validateHref(mentionHrefs, pattern);
+    }
+
+    /**
+     * @returns an object containing flags for various ways the configuation can be invalid
+     */
+    public getValidity(): RichTextMentionValidity {
+        return this.getValidationFlags();
+    }
+
+    /**
+     * Sets a particular validity condition flag's value, e.g. "hasInvalidFooValue" = true
+     */
+    protected setConditionValue(
+        name: FlagNames<typeof baseValidityFlagNames | ValidityFlagNames>,
+        isInvalid: boolean
+    ): void {
+        if (isInvalid) {
+            this.track(name);
+        } else {
+            this.untrack(name);
+        }
+        this.updateMentionInternalsFlag();
+    }
+
+    private updateMentionInternalsFlag(): void {
+        this.mentionInternals.validConfiguration = this.isValid();
+    }
+
+    private validateMissingPattern(pattern: string | undefined): void {
+        this.setConditionValue(
+            'missingPatternAttribute',
+            pattern === undefined
+        );
+    }
+
+    private validatePattern(pattern: string | undefined): void {
+        this.setConditionValue(
+            'unsupportedPatternValue',
+            pattern === undefined || this.isInvalidRegex(pattern)
+        );
+    }
+
+    private validateMappingTypes(mappings: Mapping<unknown>[]): void {
+        const valid = mappings.every(mapping => this.supportedMappingElements.some(
+            mappingClass => mapping instanceof mappingClass
+        ));
+        this.setConditionValue('unsupportedMappingType', !valid);
+    }
+
+    private validateUniqueMentionHref(
+        mentionHref: (string | undefined | unknown)[]
+    ): void {
+        const invalid = new Set(mentionHref).size !== mentionHref.length;
+        this.setConditionValue('duplicateMappingMentionHref', invalid);
+    }
+
+    private validateNoMissingMentionHref(mentionHrefs: unknown[]): void {
+        const invalid = mentionHrefs.some(href => href === undefined);
+        this.setConditionValue('missingMentionHrefValue', invalid);
+    }
+
+    private validateHref(
+        mentionHrefs: unknown[],
+        pattern: string | undefined
+    ): void {
+        if (
+            this.getValidationFlags().unsupportedPatternValue
+            || this.getValidationFlags().missingPatternAttribute
+        ) {
+            return;
+        }
+        const invalid = this.isInvalidRegex(pattern!)
+            ? true
+            : mentionHrefs.some(href => {
+                return (
+                    href === undefined
+                      || typeof href !== 'string'
+                      || this.isInvalidUrl(href)
+                      || !RegExp(pattern!).test(href)
+                );
+            });
+        this.setConditionValue('unsupportedMentionHrefValue', invalid);
+    }
```

Comment:
I find the logic of this confusing. This should be modified in two ways:
1. Don't rely on implicit validation from other code paths. Here it is assuming that some other functions have run first and set some flags. Don't access the other flags
2. I'd expect each code path to result in `unsupportedMentionHrefValue` being set. As written there are code paths that don't set the flag at all.

This is the error reporting. If I told you the error is `unsupportedMentionHrefValue` you should be able to give a succinct answer as to why that is the case. Right now it would be hard to explain.

Maybe instead of `missingMentionHrefValue` and `unsupportedMentionHrefValue` we should have `mentionHrefNotValidUrl`, i.e. it fails to parse as a URL (because it is an invalid string, such as empty string) and `mentionHrefDoesNotMatchPattern`, i.e. the URL does not match the pattern.

Reply:
Added `mentionHrefDoesNotMatchPattern` and `mentionHrefNotValidUrl` and removed `mentionHrefNotValidUrl`.